### ISSUE TITLE
Enabling RetryListenerSupport breaks @Retryable methods

### ISSUE THREAD

**Issue by bradyforcier1** [spring-projects/spring-retry#193]
I have an interface with parameters I apply to a bunch of methods in a wrapper class

    	@Retention(RetentionPolicy.RUNTIME)
	@Retryable(maxAttempts = 3,
				listeners = {"springRetryLoggingListener"},
				value = { AmazonClientException.class }, 
				backoff = @Backoff(
						delay = 1000, 
						multiplier = 2))
	public @interface AWSS3Retryable {}

I want to add a listener that will log when a retry occurs. I create this bean, and from the config above I am referencing it.

    @Component
    public class SpringRetryLoggingListener extends RetryListenerSupport {
    
    	private final static Logger LOG = LoggerFactory.getLogger(SpringRetryLoggingListener.class);
    
    	@Override
    	public <T, E extends Throwable> void onError(RetryContext context, RetryCallback<T, E> callback, Throwable throwable) {
    		LOG.warn("Retryable method {} threw {}th exception {}", context.getAttribute("context.name"), context.getRetryCount(), throwable.toString());
    		super.onError(context, callback, throwable);
    	}
    }

However, as soon as I do this, the retry stops occurring. Even if I don't reference the bean in the @Retryable block it doesn't work. It appears that even having a RetryListenerSupport bean loaded by spring it won't work. Using 1.2.5-RELEASE

Has anyone experienced this before?

Thanks

**Comment by dsyer** [comment]
Are you sure the bean name is `springRetryLoggingListener`? You haven't shown a complete sample so I'm not 100% sure what is going on.

**Comment by kaintharinder** [comment]
Hi @bradyforcier1 
remove the super part from onError.

**Comment by dsyer** [comment]
Do you have a complete sample? The OP was incomplete so it's hard to reconstruct the error.

**Comment by jaapspiering** [comment]
I have created a testcase here: https://github.com/jaapspiering/spring-retry-testcase

I am not 100% sure this is how it was set up in the OP, but at least the outcome is the same. It depends on how the RetryListener component is defined.

**Comment by dsyer** [comment]
Your comment in the configuration tells you the root of the problem:

```java
    /**
     * Breaks Retryable. If moved to a different class it works, if static it works.
     */
    @Component("uploadRetryListener")
    public class UploadRetryListener extends RetryListenerSupport { ...  }
````

By making the listener an inner (non-static) class of the bean that it is supposed to intercept, I think you have created an unresolvable cycle (no way to create the service without the listener and no way to create the listener without the service). I'm sad that there isn't an error, or at least a warning to say what is happening, but at least there is a reasonable workaround, and I'm not sure if inner (non-static) classes should ever be `@Component` really.

**Comment by jaapspiering** [comment]
Thanks for the quick response.

Indeed it looks like it ends up in some kind of unresolvable cycle. I'm not sure about defining the bean like this either, but under normal circumstances Spring (Boot) seems to support it. I could only find this comment on StackOverflow that points this out, unfortunately no official doc: https://stackoverflow.com/questions/19440525/how-to-add-inner-classes-to-spring-application-context-for-unit-testing#comment86210208_19440592

Anyway, it does work fine if the non-static inner class is placed in abean without any @Retryable annotation. It is registered and used. If defined in the same class, it must be static, otherwise it breaks, even if the listener is not actually used.

The workaround is indeed reasonable, but mainly that last point left me puzzled, I'd expect some kind of warning sign there.

**Comment by artembilan** [comment]
This works well now.
Just update https://github.com/jaapspiering/spring-retry-testcase for the latest Spring Boot version.
The output is like this:
```
open
2023-12-14T18:30:19.961-05:00 DEBUG 32508 --- [           main] o.s.retry.support.RetryTemplate          : Retry: count=0
Do a retry
onError
2023-12-14T18:30:20.963-05:00 DEBUG 32508 --- [           main] o.s.retry.support.RetryTemplate          : Checking for rethrow: count=1
2023-12-14T18:30:20.964-05:00 DEBUG 32508 --- [           main] o.s.retry.support.RetryTemplate          : Retry: count=1
Do a retry
onError
2023-12-14T18:30:21.975-05:00 DEBUG 32508 --- [           main] o.s.retry.support.RetryTemplate          : Checking for rethrow: count=2
2023-12-14T18:30:21.975-05:00 DEBUG 32508 --- [           main] o.s.retry.support.RetryTemplate          : Retry: count=2
Made it through
close
```

I believe this is because listeners are now resolved later on via `RetryConfiguration.afterSingletonsInstantiated()`.

So, marking this as `Fixed`  with respective version `2.0.0`.
